kotlinx-coroutines-core/common/src/flow/DataFlow.kt

```diff
@@ -0,0 +1,335 @@
+ * but always collects the most recently emitted value.
+ *
+ * The name reflects the fact that [DataFlow] represents an cell in a "data flow programming" model
+ * (think of an electronic spreadsheet). Dependent values can be defined by using various operators on the
+ * flows, with [combineLatest] being especially useful to combine multiple data flows using arbitrary functions.
+ *
+ * A data flow can be [closed][close] and all its collectors will complete.
+ *
+ * A read-only interface to data flow is a [Flow]. It is supposed to be collected by the consumers of the data.
+ * The ability to read the current [value] is provided only for convenience of the code that updates the data.
+ * For example, the following class encapsulates an integer data flow and increments its value on each call to `inc`:
+ *
+ * ```
+ * class CounterModel {
+ *     private val _counter = DataFlow(0) // private data flow
+ *     val counter: Flow<Int> get() = _counter // publicly exposed as a flow
+ *
+ *     fun inc() {
+ *         _counter.value++
+ *     }
+ * }
+ * ```
+ *
+ * Having two instances of the above `CounterModel` class one can define the sum of their counters like this:
+ *
+ * ```
+ * val aModel = CounterModel()
+ * val bModel = CounterModel()
+ * val sum = aModel.counter.combineLatest(bModel.counter) { a, b -> a + b }
+ * ```
+ *
+ * Conceptually data flow is similar to
+ * [ConflatedBroadcastChannel][kotlinx.coroutines.channels.ConflatedBroadcastChannel]
+ * but it is simpler, because it does not have to implement all the channel APIs.
+ *
+ * All methods of data flow are **thread-safe** and can be safely invoked from concurrent coroutines without
+ * external synchronization.
+ *
+ * ### Operator fusion
+ *
+ * Application of [flowOn], [conflate], and
+ * [buffer] with [CONFLATED][Channel.CONFLATED] or [RENDEZVOUS][Channel.RENDEZVOUS] capacity
+ * has no effect on the data flow.
+ * 
+ * ### Implementation notes
+ *
+ * Data flow implementation is optimized for memory consumption and allocation-freedom. It uses a lock to ensure
+ * thread-safety, but suspending collector coroutines are resumed outside of this lock to avoid dead-locks when
+ * using unconfined coroutines. Adding new collectors has `O(1)` amortized code, but updating a [value] has `O(N)`
+ * cost, where `N` is the number of collectors.
+ */
+@FlowPreview
+public interface DataFlow<T> : Flow<T> {
+    /**
+     * Returns `true` if this data flow is initialized.
+     *
+     * A data flow is initialized when its [value] was set at least once.
+     * A [closed][isClosed] data flow is considered to be initialized.
+     */
+    public val isInitialized: Boolean
+
+    /**
+     * Returns `true` if this data flow is closed.
+     *
+     * See [close].
+     */
+    public val isClosed: Boolean
+
+    /**
+     * Value of this data flow.
+     *
+     * Setting a value that is [equal][Any.equals] to the previous one does nothing. 
```

Comment:
Where this restriction came from? Are we sure this is desired behavior for most of the users?

Reply:
In other Reactive frameworks, such entity does not only represent `Flow`, but also `FlowCollector`.
Example of usages:
 * [Hystrix, caching example](https://github.com/Netflix/Hystrix/blob/7f5a0afc23aa5ff82320560a04d4c81a45efd67c/hystrix-core/src/main/java/com/netflix/hystrix/HystrixCachedObservable.java#L15)
 * [Kickstarted, binding subject to observable](https://github.com/kickstarter/android-oss/blob/e73c5585948b3d9435359f42a3892290c9850f6f/app/src/main/java/com/kickstarter/viewmodels/DeepLinkViewModel.java)

Note that I (currently) do not advocate for doing the same thing, but it is something definitely worth consideration (maybe @gildor can give us a few pointers here?).

Reply:
I've thought about making it implement `FlowCollector` (it is easy to add), but I could not find a complessing use-case to do it. In Kotlin word it will be less convenient, since `FlowCollector.emit` is a suspending function, do you do not actually need suspending update function. You can simply update `DataFlow.value` from any context, even non-suspending one. 

P.S. If we create some kind of `EventFlow` as a complement (which would replace `ArrayBroadcastChannel` and would not conflate events), then it would be logical to call its updating function `emit` and naturally implement `FlowCollector` interface there.

Reply:
Hi @elizarov

> FlowCollector.emit is a suspending function, do you do not actually need suspending update function

Overriding a suspending function with a non-suspending one is a current Kotlin limitation.
`DataFlow` shares this issue with `ConflatedChannel` and `UnlimitedChannel`.

On the other hand in current design to update value (`dataFlow.value=42`) is cheap, instead to read a value using the public interface is really expensive (`flow.first()`).

Reply:
Personally, I do not have good use cases for FlowCollector for DataFlow (same as rarely use io.reactivex.Observer that implemented by Subject). so my opinion that some simple adapter function that allows forwarding events from Flow/Channel/Lambda to DataFlow would be enough

Reply:
+1 for allowing same value to be set twice. If someone wants to dedup it, they can as well use `distinctUntilChanged`.
One use case for dispatching the same value is using it for UI actions, like a refresh.

```
private val refreshTrigger = DataFlow<Unit>().also {
   it.value = Unit
}
val data = refreshTrigger.flatMapLatest {
    repositroy.loadData()
}
fun refresh() { 
    refreshTrigger.value = Unit
}
```
If there is an `EventFlow`, it could also fulfill this need.